fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/join/JoinReorderHelper.java

```diff
@@ -0,0 +1,206 @@
+                containsAll.add(conjunct);
+            } else {
+                nonContainsAll.add(conjunct);
+            }
+
+        }
+    }
+
+    public static Map<ColumnRefOperator, ScalarOperator> conditionRefProjectionCol(ScalarOperator onCondition,
+                                                      Map<ColumnRefOperator, ScalarOperator> projection) {
+        Map<ColumnRefOperator, ScalarOperator> result = Maps.newHashMap();
+        if (onCondition == null) {
+            return result;
+        }
+
+        ColumnRefSet cols = new ColumnRefSet(projection.keySet());
+        ColumnRefSet onConditionColSet = onCondition.getUsedColumns().clone();
+        onConditionColSet.intersect(cols);
+
+        if (!onConditionColSet.isEmpty()) {
+            for (Map.Entry<ColumnRefOperator, ScalarOperator> entry : projection.entrySet()) {
+                if (onConditionColSet.contains(entry.getKey().getId())) {
+                    result.put(entry.getKey(), entry.getValue());
+                }
+            }
+        }
+
+        return result;
+    }
+
+    public static Map<ColumnRefOperator, ScalarOperator> mergeProject(Map<ColumnRefOperator, ScalarOperator> project,
+                                                                      OptExpression optExpression,
+                                                                      ColumnRefFactory columnRefFactory) {
+        Map<ColumnRefOperator, ScalarOperator> inputProject;
+
+        if (optExpression.getOp().getProjection() != null) {
+            inputProject = optExpression.getOp().getProjection().getColumnRefMap();
+
+        } else {
+            inputProject = optExpression.getOutputColumns().getColumnRefOperators(columnRefFactory).stream()
+                    .collect(Collectors.toMap(Function.identity(), Function.identity()));
+        }
+
+        Map<ColumnRefOperator, ScalarOperator> newProject = Maps.newHashMap(inputProject);
+
+        ReplaceColumnRefRewriter rewriter = new ReplaceColumnRefRewriter(inputProject);
+        for (Map.Entry<ColumnRefOperator, ScalarOperator> entry : project.entrySet()) {
+            newProject.put(entry.getKey(), rewriter.rewrite(entry.getValue()));
+        }
+        return newProject;
+    }
+
+    public static boolean isAssoc(OptExpression bottomJoinExpr, OptExpression topJoinExpr) {
+        LogicalJoinOperator topJoin = (LogicalJoinOperator) topJoinExpr.getOp();
+        ScalarOperator topJoinOnCondition = topJoin.getOnPredicate();
+
+        List<ColumnRefSet> splitCols = splitJoinOutputCols(bottomJoinExpr);
+
+        if (!splitCols.get(2).isEmpty()) {
+            return false;
+        } else if (topJoinOnCondition == null) {
+            return false;
+        } else if (topJoin.getJoinType() == JoinOperator.INNER_JOIN
+                && topJoinOnCondition.getUsedColumns().isIntersect(splitCols.get(1))) {
+            // when topJoin is inner join, it's on condition must ref cols from right child of bottom join
```

Comment:
The JoinAssociativityRule processes pattern like 
```
        Join                  Join
 (joinType: t1)            (joinType: t2)
      /      \               /      \
    Join      C             A       Join
(joinType: t2)     =>           (joinType: t1)
   /    \                            /    \
  A      B                          B      C
```
The topJoin on condition must contain cols from B, otherwise, it's most likely a meaningless transformation because the new bottom join is a cross join. When topJoin is inner join, the new bottom join is inner join. It's correct plan even when oldTopJoin has on condition ref cols from A like `A.v1 = C.v1`. But when top Join is other type join like `left outer join`, it will be a invalid plan when oldTopJoin has on condition ref cols from A like `A.v1 = C.v1` because the new bottom join yeilds null values of cols in C.